### PR TITLE
Move `DIAGNOSTICS` usage to `rustc_driver`

### PR DESCRIPTION
**Pull request by Centril**
Remove `rustc_interface`'s dependency on `rustc_error_codes` and centralize all usages of `DIAGNOSTICS` in `rustc_driver`. Once we remove all references to `rustc_error_codes` in all other crates but `rustc_driver`, this should allow for incremental recompilation of the compiler to be smoother when tweaking error codes. This works towards https://github.com/rust-lang/rust/issues/66210#issuecomment-551862528.

(May include traces of minor drive-by cleanup.)

r? @Mark-Simulacrum

### CODE REVIEW

File: src/librustc_driver/lib.rs [hunk 1]
```diff
@@ -187,8 +187,9 @@ pub fn run_compiler(
         config
     };
 
+    let registry = Registry::new(&rustc_error_codes::DIAGNOSTICS);
```
Comment (Mark-Simulacrum) [inline src/librustc_driver/lib.rs]: This seems to have lost the concern about not having the error codes from codegen.

File: src/librustc_interface/interface.rs [hunk 2]
```diff
@@ -143,17 +144,19 @@ pub struct Config {
         Option<fn(&Session, &mut ty::query::Providers<'_>, &mut ty::query::Providers<'_>)>,
 }
 
-pub fn run_compiler_in_existing_thread_pool<F, R>(config: Config, f: F) -> R
-where
-    F: FnOnce(&Compiler) -> R,
-{
+pub fn run_compiler_in_existing_thread_pool<R>(
+    config: Config,
+    registry: Registry,
```
Comment (Mark-Simulacrum) [inline src/librustc_interface/interface.rs]: Could this be part of `Config`? It feels a bit odd to have it separate given that it's both immutable and essentially configuration...

**Review by Mark-Simulacrum — COMMENTED** [review]
I'm not sure it's entirely worth doing this; it feels like the code is not significantly better and the improvement in compile times (2 crates -- rustc_interface and rustc_driver before, just rustc_driver now) is not really significant as neither crate is all that large.

**Comment by Mark-Simulacrum** [comment]
Modulo those nits and if you think it's worth it then r=me.

File: src/librustc_driver/lib.rs (hunk 1 above)
Comment (Centril) [reply src/librustc_driver/lib.rs]: The concern was fixed in the PR which moved the error codes.

File: src/librustc_interface/interface.rs (hunk 2 above)
Comment (Centril) [reply src/librustc_interface/interface.rs]: Sure; moved over -- this caused some slight duplication but I think it makes sense still.

**Comment by Centril** [comment]
@Mark-Simulacrum It's not a big win, but still, a bit of cleanup and a small recompilation win here and there builds to something larger. ^^

**Review by Mark-Simulacrum — APPROVED** [review]

**Comment by Centril** [comment]
Moved the macro back to avoid the problem with the index page; I'll wait with r=you for a bit.

**Pull request merged**